File: sword_offer/cpp/17_树的子结构.cpp
#include <iostream>
#include <vector>
#include <stdio.h>
#include <sys/time.h>

#include "TreeNode.h"

/*
typedef struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
} TreeNode;
*/

class Solution {
public:
    // 1. 完全内部实现另外一个递归函数
    bool HasSubtree(TreeNode* a, TreeNode* b) {
        if (a == nullptr || b == nullptr) {
            return false;
        }

        return isSubtree2(a, b, false);
    }

    // 定义用于递归调用的函数
    // 存在问题，中间会跳过节点判断，无法判断之前是否已经对齐
    bool isSubtree(TreeNode* a, TreeNode* b) {
        if (b == nullptr) {
            return true;
        }
        if (a == nullptr) {
            return false;
        }
        if (a->val == b->val) {
            return isSubtree(a->left, b->left) && \
                   isSubtree(a->right, b->right) || \
                   isSubtree(a->left, b) || \
                   isSubtree(a->right, b);
        } else {
            return isSubtree(a->left, b) || isSubtree(a->right, b);
        }
    }


    // 定义用于递归调用的函数（优化版）
    // 增加flag用来指示之前是否已经对齐（正确）
    bool isSubtree2(TreeNode* a, TreeNode* b, bool flag) {
        if (b == nullptr) {
            return true;
        }
        if (a == nullptr) {
            return false;
        }
        if (flag && a->val != b->val) { // 如果之前对齐，本次不等
            return false;
        }
        if (a->val == b->val && isSubtree2(a->left, b->left, true) && isSubtree2(a->right, b->right, true)) {
            return true;
        }
        return isSubtree2(a->left, b, false) || isSubtree2(a->right, b, false);
    }

    // 2. 自身递归即可
    // 注意题目要求空树不是任何树的子结构
    // 这一点与递归中空树是任何树的子结构是冲突的
    // 因为递归才会麻烦一些
    // 同样存在问题，无法判断之前是否已经对齐
    bool HasSubtree2(TreeNode* a, TreeNode* b) {
        if (a == nullptr || b == nullptr) {
            return false;
        }

        bool tmp = false;
        if (a->val == b->val) {
            if (b->left == nullptr && b->right == nullptr) {
                return true;
            } else if (b->left == nullptr && b->right != nullptr) {
                tmp = HasSubtree2(a->right, b->right);
            } else if (b->right == nullptr && b->left != nullptr) {
                tmp = HasSubtree2(a->left, b->left);
            } else {
                tmp = HasSubtree2(a->left, b->left) && HasSubtree2(a->right, b->right);
            }
        }

        if (tmp) {
            return tmp;
        } else {
            return HasSubtree2(a->left, b) || HasSubtree2(a->right, b);
        }
        // 等价为： return tmp || HasSubtree2(a->left, b) || HasSubtree2(a->right, b);
    }

    // 3. 两个函数，互相递归，先序遍历（根、左、右）
    bool HasSubTree3(TreeNode* a, TreeNode* b) {
        return (a != nullptr && b != nullptr) && (isSubTree3(a, b) || HasSubTree3(a->left, b) || HasSubTree3(a->right, b));
    }

    // 判断是否是子结构，辅助函数，不进行向下的延伸
    bool isSubTree3(TreeNode* a, TreeNode* b) {
        if (b == nullptr) {
            return true;
        }
        if (a == nullptr || a->val != b->val) {
            return false;
        }
        return isSubTree3(a->left, b->left) && isSubTree3(a->right, b->right);
    }

    // 测试函数
    void test(std::vector<int>& A, std::vector<int>& B) {
        bool result;
        struct timeval start, end;
        TreeNode* a = TreeNode_creatFromFrontOrder(A);
        TreeNode* b = TreeNode_creatFromFrontOrder(B);

        printf("=====\n");
        for (auto & func : this->func_vec_) {
            gettimeofday(&start, nullptr);
            result = (this->*func)(a, b);
            gettimeofday(&end, nullptr);
            printf("result: %d, times(us): %ld\n", result, end.tv_usec - start.tv_usec);
        }
    }

private:
    typedef bool (Solution::*func_ptr)(TreeNode*, TreeNode*);
    std::vector<func_ptr> func_vec_ = {&Solution::HasSubtree,
                                       &Solution::HasSubtree2,
                                       &Solution::HasSubTree3};
};


int main(int argc, char* argv[])
{
    std::vector<int> A = {8, 8, 9, -1, -1, 2, 4, -1, -1, 7, -1, -1, 7, -1, -1};
    std::vector<int> B = {8, 9, -1, -1, 2, -1, -1};
    std::vector<int> A2 = {1, 0, -4, -1, -1, -3, -1, -1, 1, -1, -1};
    std::vector<int> B2 = {1, -4, -1, -1, -1};

    Solution s;
    s.test(A, B);
    s.test(A2, B2);

    return 0;
}

File: sword_offer/python/17_树的子结构.py
#!/bin/bash python3
#-*- coding:utf-8 -*-

import time
import TreeNode

"""
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
"""

class Solution:
    def HasSubtree(self, a, b):
        """
        1. 内部定义递归函数
        注意需要传递flag记录之前是否已经对齐
        """
        def isSubTree(a, b, flag):
            if b == None:
                return True
            if a == None:
                return False
            if flag and a.val != b.val: # 如果之前对齐，本次不等
                return False
            if a.val == b.val and isSubTree(a.left, b.left, True) and isSubTree(a.right, b.right, True):
                return True
            else:
                return isSubTree(a.left, b, False) or isSubTree(a.right, b, False)
  
        # 调用子函数
        if a == None or b == None:
            return False
        return isSubTree(a, b, False) 

    def HasSubtree2(self, a, b):
        """
        2. 更加高效的方式
        """
        def isSubTree(a, b):
            if b == None:
                return True
            if a == None or a.val != b.val:
                return False
            return isSubTree(a.left, b.left) and isSubTree(a.right, b.right)
        
        return (a != None and b != None) and (isSubTree(a, b) or self.HasSubtree2(a.left, b) or self.HasSubtree2(a.right, b))
    
    def test(self, A, B):
        """
        测试函数
        """
        func_vec = [self.HasSubtree, self.HasSubtree2]
        head = TreeNode.TreeNode(0)
        a = head.creatFromFrontOrder(A)
        b = head.creatFromFrontOrder(B)

        print("=====")
        for func in func_vec:
            start = time.time()
            result = func(a, b)
            end = time.time()
            print("result: {}, times(us): {:>5.2f}".format(result, (end - start)*10**6))


def main():
    A = [8, 8, 9, -1, -1, 2, 4, -1, -1, 7, -1, -1, 7, -1, -1]
    B = [8, 9, -1, -1, 2, -1, -1]
    A2 = [1, 0, -4, -1, -1, -3, -1, -1, 1, -1, -1]
    B2 = [1, -4, -1, -1, -1]

    s = Solution()
    s.test(A, B)
    s.test(A2, B2)


if __name__ == "__main__":
    main()